Unicode Rethink [Do Not Merge]

dabrahams opened the pull request:
Lays foundation for String rethink.

The following failures are known to exist

FAIL: Swift(macosx-x86_64) :: compiler_crashers/28616-swift-parser-parseexprsequence-swift-diag-bool-bool.swift (957 of 9168)
FAIL: Swift(macosx-x86_64) :: stdlib/UnicodeTrie.swift.gyb (1048 of 9168)
FAIL: ["GraphemeClusterSegmentation/UnicodeSpec"]
FAIL: Swift(macosx-x86_64) :: stdlib/String.swift (1050 of 9168)
FAIL: Swift(macosx-x86_64) :: stdlib/InputStream.swift.gyb (5805 of 9168)
[     FAIL ] ReadLine.Unicode/strippingNewline=false
[     FAIL ] ReadLine.Unicode/strippingNewline=true
FAIL: ["Unicode/strippingNewline=false", "Unicode/strippingNewline=true"]
FAIL: Swift(macosx-x86_64) :: api-digester/source-stability.swift (5983 of 9168)
FAIL: Swift(macosx-x86_64) :: stdlib/Character.swift (7023 of 9168)
[     FAIL ] Character.RoundTripping/Random
FAIL: ["RoundTripping/Random"]
FAIL: Swift(macosx-x86_64) :: stdlib/Renames.swift (7048 of 9168)

dabrahams commented:
@swift-ci Please test Linux platform

airspeedswift commented on test/Prototypes/Unicode.swift:
```diff
@@ -49,6 +50,25 @@ extension ParsedUnicode {
 }
 
 //===--- Missing stdlib niceties ------------------------------------------===//
+postfix operator ^
```
😐

dabrahams commented on test/Prototypes/string_prototype/StringComparison.swift:
```diff
@@ -58,23 +58,88 @@ where
       //
       // TODO: cost/benefit investigation of more complex check here...
-      return codeUnit < 0xD800
+      return codeUnit < 0xD800 || codeUnit > 0xDFFF
```
Please describe the more complex check in the comment

milseman replied:
I adjusted the surrounding comments in my latest push. Sorry, I've been trying to wrap everything up today before heading out, and the history got a little chaotic at the end.

dabrahams reviewed (commented):
Very excited to see this series of commits, @milseman!

airspeedswift commented:
@swift-ci  please test linux platform

dabrahams commented on test/Prototypes/string_prototype/StringCompatibility.swift:
```diff
@@ -62,3 +72,31 @@ extension String {
   }
 }
 
+// TODO: this on the new String
```
Please use `#if false` for these kinds of things to cut down on VCS churn

airspeedswift commented on test/Prototypes/string_prototype/Substring.swift.gyb:
```diff
@@ -0,0 +1,203 @@
+  // public typealias ${ViewPrefix}Index = String.Valid${ViewPrefix}View.Index
+
+  public var unicodeScalars: String.UnicodeScalarView {
+    get {
```
is there a style/future-proofing reason to not elide the `get` here?

milseman commented on test/Prototypes/string_prototype/main.swift:
```diff
@@ -335,4 +335,15 @@ testSuite.test("substring") {
+testSuite.test("bridging") {
+  defer { _debugLogging = false }
+  _debugLogging = false
+  let s : main.String
```
Should that be: `defer { _debugLogging = true }`?

dabrahams replied:
No, the intention was to change line 341 to `= true` to turn on logging for just this case.

Commit:
[stdlib] StringProto: de-underscore _UnicodeStorage

Commit:
[stdlib] StringProto: pull out AnyUnicode[Content]

This gives us a common type to which we can downcast and use optimized storage
based on the encoding.  Use it to avoid re-flattening a character view when
possible.  This is the same mechanism we'll use to optimize pattern matching for
String content.

Commit:
[stdlib] StringProto: Unicode/AnyUnicode => stdlib

Commit:
[emacs-support] Disable flymake for Swift

It's just getting in the way for stdlib development.  Individuals can add it
back if they choose.

Commit:
[stdlib] Remove stray %end from gyb file

Commit:
[stdlib] StringProto: make views in-place RangeReplaceabble

Still needs testing, but the compiler's happy at least.

Commit:
[stdlib] StringProto: exercise inplace view mutation

Commit:
[stdlib] StringProto: add storage for arbitrary NSString

This case of the enum isn't really supported yet, but it does prove that our
String still fits in 64 bits.

Commit:
[stdlib] Add AlternateIntegerWidths.swift

so we can play with enum layout.

Commit:
[stdlib] StringProto: don't rely on spare bits bug for one-word size

See https://bugs.swift.org/browse/SR-4365.  Instead of using UnknownObject,
we'll store non-tagged NSString representations into the “Cocoa” case and
translate the tagged reps into one of our small representations.

Commit:
[stdlib] StringProto: move some defaults up to AnyUnicodeContent

Commit:
[stdlib] StringProto: Fill out support for the NSString case

Commit:
[stdlib] StringProto: minor cleanups and comments

And now, to bed.

Commit:
index on unicode-rethink: ee0f1219f05 [stdlib] StringProto: minor cleanups and comments

milseman committed:
String prototype: Move normalized view into stdlib proper

Commit:
[stdlib] Make FixedArray public

Commit:
[stdlib] StringProto: (un)pack bits for small string representation

Commit:
[stdlib] StringProto: small string optimization

milseman committed:
String prototype: Switch to using the new FCCNormalizedUTF16View.

Hook up the new lazy segmented normalized view, and use that inside of
_UnicodeViews.

Commit:
Merge remote-tracking branch 'origin/master' into unicode-rethink

milseman committed:
Merge pull request #8399 from milseman/wip_strings_prototype

String prototype: Switch to using the new FCCNormalizedUTF16View.

Commit:
[stdlib] StringProto: drag some tests over from main.swift

Commit:
[stdlib] StringProto: drag more tests over from main.swift

printing/character/hashable/comparable

Commit:
[stdlib] StringProto: Cstring support and  tests

Commit:
Merge Michael's changes

Commit:
[stdlib] StringProto: Bridging support and  tests

Commit:
[stdlib] StringProto: Better bridging tests

milseman committed:
String prototype: New String has _StringCore by shuttling through old String

milseman committed:
Merge pull request #8412 from milseman/wip_strings_prototype

String prototype: New String has _StringCore by shuttling through old…

Commit:
[stdlib] StringProto: Avoid creating multiple layers of type-erasure

Commit:
[stdlib] StringProto: Bypass needless type-erasure

Commit:
Merge my work

Commit:
[stdlib] StringProto: Forward CharacterView through .any case

Commit:
[stdlib] StringProto: SubString type

Doesn't yet conform to Unicode

milseman committed:
String prototype: get off of old String by bridging for _StringCore

milseman committed:
Merge pull request #8436 from milseman/wip_strings_prototype

String prototype: get off of old String by bridging for _StringCore

milseman committed:
WIP: BROKEN: starting integration

milseman committed:
WIP: BROKEN: continuing integration

milseman committed:
WIP: BROKEN: whoops, include the new file

milseman committed:
WIP: BROKEN: Swift.o compiles!!!!1one!

milseman committed:
WIP: Try sprinkling some numericCasts around

Commit:
[stdlib] StringProto: fix O(N^2) growth bug

Commit:
[stdlib] StringProto: fix O(N^2) growth bug

milseman committed:
WIP: still broken: make more things public

milseman committed:
WIP: hack Foundation's (probably wrong) Index interchange

milseman committed:
WIP: keep StdlibUnittest putting along

Commit:
[stdlib] StringProto: properly forward _tryToReplaceSubrange

Forwarding mutating methods implies base must be a get/set property, and
knock-on effects from there.

Commit:
[stdlib] StringProto: properly forward _tryToReplaceSubrange

Forwarding mutating methods implies base must be a get/set property, and
knock-on effects from there.

Commit:
[stdlib] StringProto: Substring BOOYAH

That was really hard, not least because of the TranscodedView encodedOffset
translation bug I fixed in 29803099ce2.  That was two days of abject torture.

milseman committed:
WIP: Don't be dumb; don't import Darwin. It builds now!

milseman committed:
WIP: Manually apply 117b78c3 to NewString.swift.gyb

milseman committed:
WIP: Disable the old old prototype

milseman committed:
WIP: comment cleanup and test fixups

milseman committed:
WIP: Whoops, infinite recursion

milseman committed:
WIP: Fix crash assuming _StringCore

milseman committed:
WIP: Add scratch prototype test to try some things out

milseman committed:
WIP: Beef up the integration proto; for quick testing

milseman committed:
WIP: bypass more access control, for ease of quick hacking

milseman committed:
WIP: Bug fix, single scalars are not FCC pre-normalized

Commit:
[stdlib] StringProto: fix TranscodedView encodedOffset translation

Commit:
Doing our best to keep current with the latest work in master

airspeedswift committed:
temporarily hack two String benchmarks to work... needs to be unwound

milseman committed:
WIP: Use Int for IndexDistnace for now, otherwise many tests are broken

airspeedswift committed:
The CChar is a lie

airspeedswift commented on test/Prototypes/AU3.swift.gyb:
```diff
@@ -814,6 +852,70 @@ extension String : Hashable {
+    var result = ContiguousArray<CChar>()
+    let source = Counted(content.utf8)
+    result.reserveCapacity(numericCast(source.count + 1))
+    result.append(contentsOf: source.lazy.map { numericCast($0) })
```
Doesn't this need to be `Int8.init(bitPattern:)` rather than `numericCast`, which will trap for UTF8 values > 128?

Commit:
[stdlib] StringProto: Sort out uniqueness checking trix

Move things back into the stdlib where possible.  Restore internal-ness of
_isUnique.

Commit:
[stdlib] Add an explicit @nonobjc mark

airspeedswift committed:
remove closure from utf8.map

milseman committed:
WIP: Switch Character over to _UTF16StringStorage; fix inf recursion

milseman committed:
Merge remote-tracking branch 'origin/unicode-rethink'

milseman committed:
WIP: Incorporate Dave's latest changes

milseman committed:
Merge remote-tracking branch 'milseman/wip_strings_prototype_integration'

Commit:
cmake: Teach cmake that gyb files depend on foo in ``gyb.expand('foo.swift.gyb'...)``

Fixes https://bugs.swift.org/browse/SR-4366.
Fixes rdar://problem/31331437.

erg committed:
Merge pull request #8492 from erg/cmake-gyb-expand-unicode-rethink

Cmake gyb expand unicode rethink

milseman committed:
WIP: Integration: remove commented out old code

milseman committed:
Merge pull request #8501 from milseman/wip_strings_prototype_integration

WIP: New string integrated into stdlib

Commit:
[stdlib] StringProto: remove flotsam

this method was relocated to Builtin.swift

milseman committed:
String prototype: Prevent potential circularity in description

Commit:
[stdlib] StringProto: clean up a bit of cString interop

milseman committed:
String prototype: add length() selector to EmptyArray, for NSString...

milseman committed:
Merge pull request #8520 from milseman/wip_strings_prototype

String prototype: 2 small bug fixes

milseman committed:
String prototype: EmptyArray has more stringy selectors

milseman committed:
String prototype: remove more unavailable APIs, that may be currently available

milseman committed:
Merge pull request #8524 from milseman/wip_strings_prototype_integration

Strings Prototype: squash 2 crashers

milseman committed:
String prototype: Remove more unavailable APIs

milseman committed:
String prototype: Drop more unavailable functions; dynamically chosen sometimes

milseman committed:
Merge pull request #8527 from milseman/wip_strings_prototype_integration

String Prototype: Drop more unavailable APIs

milseman committed:
String prototype: Substring is Comparable, and [Debug]StringConvertible

milseman committed:
String prototype: change/disable bunch of tests to accomodate non-nil BufferPointer baseAddress

milseman committed:
String prototype: Sprinkle lots of String() constructors from Substring. More coming

milseman committed:
Merge pull request #8536 from milseman/wip_strings_prototype_integration

String prototype: Fix up a bunch of test suite compilation errors

Commit:
[stdlib] Forward collection APIs without index wrap/unwrap

...at least make it possible.

milseman committed:
String prototype: Quick note that CR-LF is a single grapheme

Commit:
Move AU3 test into StringIntegration

airspeedswift committed:
Revert non-optional unsafe buffer base address (#8542)

* revert removal of baseAddress optionality

* revert second commit

* Revert "String prototype: change/disable bunch of tests to accomodate non-nil BufferPointer baseAddress"

This reverts commit abc5f2804fa148c9eca12211016c7f0f79f13a45.

* Update new code to handle nullable unsafe buffer

Commit:
Merge remote-tracking branch 'origin/unicode-rethink' into unicode-rethink

Commit:
[stdlib] StringProto: restore bridging of small strings

Commit:
[stdlib] StringProto: Add closure-less _MapCollection

pruning sources of overhead

milseman committed:
String prototype: Try out a Latin1CharacterView; Will probably rename/redesign soon

milseman committed:
Merge pull request #8565 from milseman/wip_strings_prototype

String prototype: Try out a Latin1CharacterView; Will probably rename/redesign soon

milseman committed:
WIP: Whoops. Remove my print lines...

milseman committed:
Merge pull request #8566 from milseman/wip_strings_prototype

WIP: Whoops. Remove my print lines...

milseman committed:
String prototype: Add integration-prototype test for CR-LF single grapheme

milseman committed:
Merge pull request #8567 from milseman/wip_strings_prototype

String prototype: Add integration-prototype test for CR-LF single gra…

Commit:
[stdlib] Add missing public

Commit:
Merge remote-tracking branch 'origin/unicode-rethink' into unicode-rethink

Commit:
[stdlib] StringProto: Eliminate existential from String.Index

milseman committed:
String prototype: update subString.swift test; minor change to top-level utf16 view

milseman committed:
Merge pull request #8570 from milseman/wip_strings_prototype

String prototype: update subString.swift test; minor change to top-le…

milseman committed:
String prototype: Use _UTF16StringStorage for empty String isa pointer, not _EmptyArrayStorage

milseman committed:
Merge pull request #8575 from milseman/wip_strings_prototype

String prototype: Use _UTF16StringStorage for empty String isa pointe…

airspeedswift committed:
Added iffed-out skeleton of a Swift 3 compatibility shim for Substring

Commit:
[stdlib] StringProto: Revamp Character, fix its test

simplify; improve storage efficiency.  Should run much faster now.

milseman committed:
String prototype: restore force unwraps of UnsafeBufferPointer base address

milseman committed:
String prototype: Gut now-irrelevant tests. We should replace with relevant ones

milseman committed:
Merge pull request #8586 from milseman/wip_strings_prototype

String prototype: gutting irrelevant tests

milseman committed:
String prototype: Unavailable APIs are irrelevant now, because now we have them

milseman committed:
String prototype: Update test case, as CharacterView's SubSequence type changed

milseman committed:
String prototype: Add explicit String constructor around a Substring

milseman committed:
Merge pull request #8592 from milseman/wip_strings_prototype

String prototype: Fix up more tests

milseman committed:
Revert "TEMPORARY: Revert "Verify the construction of SubstitutionMaps""

This reverts commit 3d546fc1a57340135227727e633b56043345f10b.

milseman committed:
Merge remote-tracking branch 'origin'

milseman committed:
Merge remote-tracking branch 'origin'

milseman committed:
Unicode rethink: Revert SubstitutionMap.cpp to origin/master

milseman committed:
Merge pull request #8603 from milseman/wip_strings_prototype

Unicode rethink: merge in master

milseman committed:
String prototype: Delete commented out \#unavailable APIs. Remove now-irrelevant tests

milseman committed:
Merge pull request #8612 from milseman/wip_strings_prototype_pr

String prototype: Delete commented out \#unavailable APIs. Remove now…

Commit:
[stdlib] StringProto: Actually use the Latin1 header

Before, all strings were getting the UTF16 header.

Commit:
[stdlib] StringProto: fix up the StringReallocation test

It still fails because string append is going through _core rather than using
RangeReplaceabble conformance.

milseman committed:
Merge remote-tracking branch 'origin'

milseman committed:
Merge pull request #8624 from milseman/wip_strings_prototype

Unicode rethink: merge in bug fixes from master

Commit:
[stdlib] StringProto: Different uniqueness checking scheme

Probably unneeded after we get upstream optimizer fixes.

Commit:
[stdlib] Expand ForwardCollectionAPIs to be able to switch

milseman committed:
Unicode rethink: Sprinkle explicit String constructors around to make the types happy

milseman committed:
Unicode rethink: XFAIL NewString.swift, as it tests internals of _StringCore

Commit:
[stdlib] ForwardCollectionAPIs: support nonmutating-only

Commit:
[stdlib] StringProto: Dispatch character accesss more efficiently

The other views could use the same treatment.

milseman committed:
Merge pull request #8637 from milseman/wip_strings_prototype

Unicode rethink: Pass test NewString.swift (which is really an old string)

Commit:
Revert "[stdlib] StringProto: Different uniqueness checking scheme"

This reverts commit 3518985c51923371bcfe33bd9e44b9305cd0d9d9; as suspected
upstream optimizer fixes made the change unnecessary.

Commit:
[stdlib] StringProto: speculatively optimize s1.append(s2)

Also leave some comments about "tricky" code.

Commit:
[stdlib] StringProto: get StringIntegration test compiling again.

Lots of failures still; needs attention.  All of the tests I put in this file
should really have gone somewhere else.  Once they're working again they should
be transplanted.

Commit:
[stdlib] StringProto: Fix a bug in grapheme breaking

Commit:
[stdlib] StringProto: Fix transcoded view index remapping

and with it, the StringIntegration test.

Commit:
[stdlib] StringProto: Fix the print_stdlib test

public-but-hidden APIs need to stay that way

Commit:
[stdlib] StringProto: Test new emoji splitting in CharacterView

milseman committed:
Unicode rethink: hashValue uses switch-dispatch; 40x speedup.

By using switch dispatch, we can access the FCC normalized UTF16
without type erasure for many of our content kinds. This results in
generic specialization for many of our collection wrappers, allowing
for a ~40x speedup for e.g. Latin1Strings.

Commit:
[stdlib] StringProto: Move some prototype tests into test/stdlib

Commit:
[stdlib] _UnicodeViews: loosen constraints

The inability to have a transcoded view of bidirectional code units was cramping
my style.

Commit:
[stdlib] StringProto: de-fragilize a test

Commit:
[stdlib] More inlining for lazy flatten collections

milseman committed:
Unicode rethink: comparison uses switch-dispatch; 40x speedup.

Similar to hashValue, doing the NxN switch and dispatch yields a
20-40x speedup over using the UnyUInt16View.

Commit:
[stdlib] Re-alphabetize CMakeLists.txt

Commit:
[stdlib] StringProto: consolidate related code...

...in preparation for more gyb'ing

Commit:
[stdlib] StringProto: de-fragilize typealiases

We're just forwarding from Content, so don't nail down the type here.

Commit:
[stdlib] Strip out string view printing / reflection

This stuff gets in the way of debugging.  We can bring stuff like this back
selectively when we get to dealing with the debugger.

Commit:
[stdlib] StringProto: Faster string view dispatch

milseman committed:
Unicode rethink: Quick pre-normalization check.

Adds a quick pre-normalization check that can skip calling over to ICU
on obviously pre-normalized sequences.

milseman committed:
Merge pull request #8654 from milseman/wip_strings_prototype

Unicode rethink: Switch dispatch more, speedups.

Commit:
[stdlib] StringProto: fix up the RenewString.swift test

Got to kill off the rest of the test failures ASAP or regressions like this will
continue to be overlooked.

Commit:
[stdlib] StringProto: fix Latin-1 -> UTF-8 transcoding

Commit:
[stdlib] Handle indices leaping off the ends of Collections

This might stop some infinite loops in the string prototype

Commit:
[stdlib] StringProto: reencode fccNormalizedUTF16 indices for String

The old encoding was losing the inner offset and thus failing to advance in any
segment of length more than 1

milseman committed:
Unicode rethink: Check inner offset for == and <

milseman committed:
Unicode rethink: Use FCCNormalizedSegment in small string; 10x speedup

This makes FCCNormalizedSegment be a UnicodeView. We can then use it
for the fccNormalizedUTF16 property on the small string types, rather
than a Any construct. This yields a 10x perf improvement for comparing
small Latin1 strings, and likely much more for full unicode strings.

milseman committed:
Merge pull request #8704 from milseman/wip_strings_prototype

Unicode rethink: Use FCCNormSegment for small string; 10x speedup

airspeedswift committed:
Remove reference to old internals from String(repeating:count)

airspeedswift committed:
Remove redudant String.isEmpty implementation

airspeedswift committed:
excise unused _split implementations

milseman committed:
Unicode rethink: "Explode" the packed packed code units; ~2x speedup.

This introduces a hard-coded fixed-size array, and de-compresses the
small string representations into it. This yields an easy 2x gain for
small latin1 string comparisons, and enables more future improvements.

Hard-coded to avoid exponential generic substitutions in release
builds, or exponential perf in debug builds.

Commit:
[stdlib] ForwardCollectionAPIs: basic bounds checking support

This only checks for out-of-bounds subscripts, but it appears to be enough for
the string tests... for now.

Commit:
[stdlib] StringProto: add basic bounds checking to views

Commit:
[stdlib] StringProto: use BidirectionalCollectionWrapper for String/Substring

Now that the views are optimized with switch dispatch and have bounds checking,
it's much better for String and Substring to avoid explicitly forwarding.

Commit:
[stdlib] StringProto: restore view reflection customizations

Turns out we need this stuff in order to pass tests.  To introspect the details
of any view, dump its `content` property.

Commit:
Use line-directive to process the output of gyb'd tests

Also make the result non-editable in emacs at least, so we don't edit the output
by mistake.

Commit:
[stdlib] StringProto: implement hasPrefix/Suffix correctly, portably

...if not yet efficiently.  We need to add the logic to make sure the end of the
putative prefix/suffix starts and ends on grapheme cluster boundaries without
scanning all the characters.

Commit:
[stdlib] StringProto: fix a bug in reverse character indexing

Commit:
[stdlib] StringProto: fix a bug in Latin1CharacterView

Commit:
Update a test for new string

because pestering people to remove stdlib dependencies is a pain.

Commit:
XFAIL tests that seem to improperly depend on String details

These tests are broken by the new String design and don't seem to be intending
to test anything string-related.

Commit:
Remove improper stdlib dependencies from a couple tests

In these cases it was pretty easy to declare a dummy type that has the old
String's layout

Commit:
[stdlib] StringProto: move legacy tests up into their own file

...discarding all the obsoleted support code that made them work.  They need to
be adapted to the new String design or discarded.

Commit:
[stdlib] StringProto: adjust a bounds check test

Somehow, the old String failed to detect when you decremented its UTF16View
index past startIndex, but the new one traps a little earlier.

Commit:
[stdlib] StringProto: adjust a test for new comparison semantics

Now Strings sort by lexicographical comparison of FCC-normalized forms.

Commit:
[stdlib] StringProto: adjust a test to avoid tagged pointer strings

...which don't have the usual lifetime relationship with associated objects.

Commit:
Squash some noisy warnings in a test

Commit:
XFAIL tests due to https://bugs.swift.org/browse/SR-4565

Commit:
XFAIL tests that fail with new String API for understandable reasons

It's not clear how to update these tests, but they aren't actually testing
anything crucial to String.

Commit:
Adjust a test for the addition of postfix '...'

       x...-y

is now ambiguous instead of just demanding spaces around the '...'

Commit:
XFAIL tests that are mystery failures on the unicode-rethink branch

We don't yet understand why these are failing

Commit:
Merge fixes for existing (non-validation) test failures.

Commit:
One more mystery XFAIL

<rdar://problem/31566154> Pure virtual function called

Commit:
[stdlib] StringProto: comparison works better; Drop an XFAIL or 2

Commit:
Merge remote-tracking branch 'origin/master' into unicode-rethink

milseman committed:
Unicode rethink: Start Specializing some views 5or6 vs 7or16

milseman committed:
Unicode rethink: Custom append for UnboundedCapacity; 20% speedup.

Add a customization hook for append for UnboundCapacity, which should
be super trivial when in the small form. This saves us an easy ~20% on
small-string comparison benchmark, in which we rarely overflow the
bounds. Should similarly improve the FCCNormalizationSegment.

Commit:
Squash warnings in tests

It's really hard to analyze failures amongst all the spew.

Commit:
Gyb'd tests should use line-directive

People, please, for the love of Mike.

milseman committed:
Unicode rethink: Trap is guaranteed on access, not necessarily on index advancement

milseman committed:
Unicode rethink: Quick bug fix, we can have empty segments

milseman committed:
Merge pull request #8730 from milseman/wip_strings_prototype_pr

Unicode rethink: explode code units; 3x speedup

Commit:
[stdlib] StringProto: make some stuff public

...for consistency and testing purposes if nothing else

airspeedswift committed:
Add very simple benchmarks for testing String

airspeedswift committed:
Update stringbench.swift

Commit:
[stdlib] Fix some UnsafeRawBufferPointer formatting issues

Commit:
[stdlib] Improve compile times by breaking up some expressions

and removing numericCasts.

Commit:
[stdlib] StringProto: add some missing @nonobjc

Commit:
[stdlib] StringProto: Replace LazyMapCollection with _MapCollection

The former holds a closure and is harder to optimize.

There are two more opportunities to do this, but trying to do one of
them (<https://github.com/dabrahams/swift/commit/e6b721cad13>)
induced https://bugs.swift.org/browse/SR-4577.

Commit:
[stdlib] Make some common types conform to ContiguouslyStored.

This API may need to be on Sequence instead so that things like _copyContents
can use it.

Commit:
[stdlib] Fix docs for _copyContents

Commit:
Merge remote-tracking branch 'origin/unicode-rethink' into unicode-rethink

Commit:
[stdlib] StringProto: 4.6x speedup on replaceRange

Special-case replaceSubrange(:with: String)

milseman committed:
Unicode rethink: Update fixed array to use buffer pointers; NFC

Commit:
[stdlib] StringProto: 1.3x speedup on replaceSubrange

Reformulated _BoundedBufferReference's replaceSubrange implementation so that
uses _copyContents(initializing:) and the new _copyContents(assigning:).

Commit:
Apply Doug's workaround for https://bugs.swift.org/browse/SR-4565

milseman committed:
Merge pull request #8755 from milseman/wip_strings_prototype

Unicode rethink: Update fixed array to use buffer pointers; NFC

Commit:
*Properly* silence test warnings.

Commit:
[stdlib] StringProto: fix a major bug in short string handling

Silly rabbit, if you only take the address of the first element, the optimizer
can throw out the rest of it.

Commit:
XFAIL a test we're not going to fix right now

misaligned indices are a low priority

Commit:
Merge remote-tracking branch 'origin/unicode-rethink' into unicode-rethink

Commit:
Merge from 'origin/master'

milseman committed:
Unicode rethink: gyb up and split out alternative FixedArray

Adds FixedArray2.swift.gyb, a potential competitor to
FixedArray.swift. This fixed array explores using gyb rather than
composition+specialization+inlining to expose a fixed size array of
elements. Results in quicker compilation times and runtimes.

milseman committed:
Unicode rethink: define iterator for Packed; 2x speedup.

This defines a quick iterator for PackedUnisignedIntegers, which nets
us a nice 2x speedup in small string comparisons (and should yield
similar boosts to small Characters).

milseman committed:
Unicode rethink: Bug fix in overly conservative pre-norm check

milseman committed:
Merge pull request #8768 from milseman/wip_strings_prototype

Unicode rethink: walk bits way faster; 2x smallstr compare

milseman committed:
Unicode rethink: _unpack has been replaced by explosions

Commit:
Suppress redundant conformance warnings

milseman committed:
Unicode rethink: Uglify ==; big speedups.

Begin uglifying String's ==. Every same-representation case has
opportunities for speedup (and, actually, so does many
different-representations).

This yields us a nice 50x speedup for == when comparing small strings
with the same small representation and the same bitsPerComponent. It
also will serve as a basis for more special-case checking.

It is very ugly, and will be getting uglier, but once we meet perf
goals I'll look into prettying it up without sacrificing perf.

Commit:
[stdlib] Uniquify duplicated test

Commit:
[stdlib] StringProto: add test for misaligned slice boundaries

milseman committed:
Unicode rethink: bug fix in normalization.

I had an overeager pre-normalization check. Make it less eager. The
pre-norm was correct for single-code-unit normalization segments, but
not for multi-code-unit segments. Fix this by making the the
multi-code-unit segment checks more conservative.

milseman committed:
Merge pull request #8783 from milseman/wip_strings_prototype

Unicode rethink: Compare bits, they're faster

swiftix committed:
Add a functionality to produce canonical conformances and substitutions

A canonical conformance is defined as a conformance which:
- does not contain any non-canonical types.
- Its type and interface type should be canonical.
- Any referenced conformances should be canonical.
- Any used substitutions should be canonical as well.

A substitution is canonical if:
- its replacement type is canonical
- all of its conformances are canonical

swiftix committed:
Fix a bug with SILBoxType’s generic arguments

The set of SILBoxType’s generic arguments should be canonical, otherwise getCanonicalType() produces wrong results.

It does not contain a reduced test-case, because the actual bug happens when compiling the stdlib on one of the experimental branches and requires a rather complex setup. I'll try to provide a test-case later.

Resolves one of the issues in rdar://31603112

swiftix committed:
Address review comments

Commit:
Revert "Apply Doug's workaround for https://bugs.swift.org/browse/SR-4565"

This reverts commit f966a8e62a5fe3cfda0656eb3b9764bdbbba08b1.

Commit:
Merge remote-tracking branch 'origin/master' into unicode-rethink

Commit:
Merge Roman's optimization performance fixups

He says this improves compile times of the optimized standard library.
Optimized builds of our benchmark still take roughly 14.5s.

Commit:
Restore the part of Doug's SR-4565 hack we still need; remove XFAILs

Commit:
Kill redundant capacity reservation

Commit:
[stdlib] RangeReplaceableCollection: static +

Commit:
Clean up an XFAIL by making it conditional

milseman committed:
Unicode rethink: No seriously, compare bits, they're way fast

Uglify small string comparison a bit more and try to reduce / remove
branches as much as possible for those paths. Finally puts us faster
than Swift 3 string for small string nano-benchmarking. This is the
first light at the end of the perf tunnel for our multi-representation
with small strings scheme.

Hopefully this can be made less ugly without sacrificing perf. Worst
case, we can #if around for any platform differences.

Commit:
[stdlib] StringProto: Remove loops from inline representation

count and subscript now do it without any loops or branches.  No significant
speedup :(

Commit:
[stdlib] StringProto: Remove .any case for 2x win on appendLong

This case is not required for parity with the existing String and we aren't
currently testing it.  We can bring it back if/when needed.

milseman committed:
Merge pull request #8808 from milseman/wip_strings_prototype

Unicode rethink: No seriously, compare bits, they're way fast

milseman committed:
Revert "Unicode rethink: No seriously, compare bits, they're way fast"

This reverts commit 029760eabe8ddafda4a8d6860fd9b4da70763d4f.

Reverted until I can come up with a less fragile scheme.

milseman committed:
Merge pull request #8816 from milseman/wip_strings_prototype

Revert "Unicode rethink: No seriously, compare bits, they're way fast"

Commit:
[stdlib] withExistingUnsafeBuffer => Sequence, forwarded by Slice

Hoist this method up to Sequence, and have slices of RandomAccessCollections
present their contiguous storage when the base doex.

Commit:
[stdlib] @inline(__always) IndexingIterator (20% speedup)

Commit:
Merge remote-tracking branch 'origin/unicode-rethink' into unicode-rethink

Commit:
[stdlib] StringProto: switch on source of growing buffer (6.6x speedup)

Pasting the instruments output into a real editor so you can actually read
what's there makes all the difference!

shahmishal commented:
@dabrahams What is the status of this PR?

dabrahams commented:
> On Oct 2, 2018, at 10:44 AM, Mishal Shah <notifications@github.com> wrote:
> 
> @dabrahams <https://github.com/dabrahams> What is the status of this PR?
> 
> 

I'm not doing anything with it; feel free to close if it's causing clutter.
